optuna/batch/trial.py

```diff
@@ -0,0 +1,154 @@
+import abc
+import datetime
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+from typing import Union
+import warnings
+
+import numpy as np
+
+import optuna
+
+
+class BaseBatchTrial(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def _get_trials(
+        self,
+    ) -> Union[
+        Sequence["optuna.trial.Trial"],
+        Sequence["optuna.multi_objective.trial.MultiObjectiveTrial"],
+    ]:
+        raise NotImplementedError
+
+    def suggest_float(
+        self,
+        name: str,
+        low: float,
+        high: float,
+        *,
+        step: Optional[float] = None,
+        log: bool = False
+    ) -> np.ndarray:
+        return np.array(
+            [t.suggest_float(name, low, high, step=step, log=log) for t in self._get_trials()]
+        )
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> np.ndarray:
+        return np.array([t.suggest_uniform(name, low, high) for t in self._get_trials()])
```

Comment:
Since trials are suggested sequentially, it's not clear from this PR how a sampler would go about implementing batched sampling. How about clarifying that, how samplers are supposed to handle batched sampling. Did you imagine that samplers should be distinguished by API or logic depending on if they're "batched samplers" or not somehow? I believe the sampler could query the batch size from the study and try to compute the batch number from the trial number, but it feels a bit complicated and might not work if trial numbers aren't in sequence within a batch (I left another comment addressing this).

We should also make sure that all trials in a batch are suggested under same circumstances. This might be guaranteed by how we call `read_trials_from_remote_storage` but I'm not sure.

Reply:
There is a trade-off between complicating the `suggest_*` logic of the `BatchTrial` and making the API of the `BatchSampler` different from that of a regular sampler, or complicating the internal logic of the `BatchSampler`. I prefer the latter.

If we choose the former, there are simply more areas that need to be fixed. Furthermore, I think there will be difficulties in integrating the `BatchTrial` with the regular `Trial` in the future.

On the other hand, if we choose the latter, the only thing we need to work on in this regard is the `BatchSampler` we'll implement in the future. Furthermore, the integration of `BatchTrial` and `Trial` would be possible without having to think about `BatchSampler`.

Therefore, I think the current implementation of `suggest_*` of `BatchTrial` is fine as it is

Reply:
Even with the current `suggest_*` logic, we can still achieve batch sampling. The `BatchSampler` samples the number of batches and stores them internally in a queue, and then it can pop them at every `suggest_*`.

Reply:
@hvy @HideakiImamura Thank you for your insightful discussion.

> How about clarifying that, how samplers are supposed to handle batched sampling.

> Even with the current `suggest_* logic`, we can still achieve batch sampling.

I agree. Indeed, CMA-ES is a kind of batch algorithm, but we have implemented it without `BatchSampler`. And I chose this design to minimize the code change.

> I believe the sampler could query the batch size from the study and try to compute the batch number from the trial number

In terms of specification of the batch size, I think users can set it when they initialize the samplers in typical usages. For example, the population size of the CMA-ES is set at the initialization. So, I think the batch sampling is supported by the current Optuna, and I'd like to focus on the batch evaluation in this PR.

> We should also make sure that all trials in a batch are suggested under same circumstances. This might be guaranteed by how we call `read_trials_from_remote_storage` but I'm not sure.

Currently, I call it right before the creation of the first trial in a batch. I don't call it when I create second and later trials, and I think the sampler can get the same observation regarding the trials not included in the batch.

As for the trials in the same batch, it depends on the samplers' algorithms. For example, TPE with the constant liar will tentatively set the constant objective values (lies) to the previous trials to prevent the sampler from suggesting the same/similar parameter values.